[MRG + 2] Rename scorers like `mse` to `neg_mse`

betatim opened the pull request:
#### Reference Issue

Fixes #2439
#### What does this implement/fix? Explain your changes.

Renaming scorers for which smaller is better (like MSE) to `neg_mse` so that they fit the idea of "bigger is better".
- [x] `mean_squared_error`
- [x] `mean_absolute_error`
- [x] `log_loss`
- [x] `median_absolute_error`

betatim committed:
Rename smaller-is-better scoring metrics

Rename scorers like MSE to neg_MSE so that it is less
surprising that they return negative values.

betatim committed:
Update tests to new scorer names

betatim committed:
Introduce deprecation warning and fix tests

get_scrorer now warns if you use an old name for a scorer
and tests have been updated to use new naming convention.

ogrisel commented on sklearn/metrics/scorer.py:
```diff
@@ -313,12 +322,12 @@ def make_scorer(score_func, greater_is_better=True, needs_proba=False,
-mean_absolute_error_scorer = make_scorer(mean_absolute_error,
-                                         greater_is_better=False)
-median_absolute_error_scorer = make_scorer(median_absolute_error,
-                                           greater_is_better=False)
```
Because those variables do not have a leading `_` the can unfortunately be considered public API even if not in the documentation. I think we should maintain backward compat aliases until 0.20:

```python
# Backward compat alias to keep until the end of the deprecation period (0.20)
mean_squared_error_scorer = neg_mean_squared_error_scorer
mean_absolute_error_scorer = neg_mean_absolute_error_scorer
median_absolute_error_scorer = neg_median_absolute_error_scorer
```

ogrisel replied:
Maybe we could add a `deprecation_message` option to `make_scorer` to make sure that the deprecation warning is raised when the user actually call the scorer instead of raising the warning in `get_scorer` that only covers a subset of the public API.

```python
neg_median_absolute_error_scorer = make_scorer(median_absolute_error,
                                               greater_is_better=False)
median_absolute_error_scorer = make_scorer(
    median_absolute_error,
    greater_is_better=False,
    deprecation_msg='Scoring method median_absolute_error was renamed to '
                    'neg_median_absolute_error in version 0.18 and will be '
                    'removed in 0.20.'
)
```

amueller replied:
why not ``deprecate(make_scorer(median_absolute_error, greater_is_better=False))`` ?

amueller replied:
Ah, it's an object, not a function, so it doesn't go into the right branch in deprecated, right? We could still wrap the "__call__" method, right?

amueller replied:
``deprecated("message")._decorate_fun(make_scorer(median_absolute_error, greater_is_better=False))``
seems to be the right solution imho.

betatim replied:
Neither of the `deprecated()._decorate_*` work because `make_scorer` doesn't return a class nor a function :-/ To get the `__name__` of a class instance you have to do `obj.__class__.__name__`.

I like handling the deprecation stuff as a wrapper instead of extra argument. I made a first proposal in 09cd4c7. It still has a few problems but it would allow you to do:

`deprecated('hello')(make_scorer(median_absolute_error, greater_is_better=False))`

This will however print `Thing _PredictScorer in deprecated...` and I think it will be quite hard to find out the correct name to print in a generic way inside `deprecated()` ... but do we want to cook something that is special to handling `make_scorer`??

ogrisel replied:
I would rather avoid making the `sklearn.utils.deprecation.deprecated` wrapper too complex. I think it's good enough to not use it at all in this case and just use a `_deprecation_msg` attribute on `_BaseScorer` as suggested in https://github.com/scikit-learn/scikit-learn/pull/7261#discussion_r76979585.

amueller replied:
I wouldn't want to make the wrapper to complex, but I feel that deprecating the call method is really the easiest thing. We also don't need to change the wrapper for that. It will then say ``__call__`` is deprecated and we can let the message explain that.

ogrisel commented on sklearn/metrics/scorer.py:
```diff
@@ -333,19 +342,19 @@ def make_scorer(score_func, greater_is_better=True, needs_proba=False,
 
 # Score function for probabilistic classification
-log_loss_scorer = make_scorer(log_loss, greater_is_better=False,
-                              needs_proba=True)
```
Same remark here.

ogrisel commented on sklearn/metrics/scorer.py:
```diff
@@ -333,19 +342,19 @@ def make_scorer(score_func, greater_is_better=True, needs_proba=False,
                average_precision=average_precision_scorer,
-               log_loss=log_loss_scorer,
+               neg_log_loss=neg_log_loss_scorer,
                adjusted_rand_score=adjusted_rand_scorer)
```
We should read the deprecated keys in the `SCORERS` dict as well. Hoping that our users will rather use `get_scorer` to get the deprecation message.

amueller replied:
read?

betatim replied:
We translated read -> leave at the sprint

ogrisel commented on sklearn/tests/test_cross_validation.py:
```diff
@@ -906,7 +906,7 @@ def test_cross_val_score_with_score_func_regression():
 
     # Mean squared error; this is a loss function, so "scores" are negative
     mse_scores = cval.cross_val_score(reg, X, y, cv=5,
```
Please rename to `neg_mse_scores` or multiply by `-1`.

ogrisel commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -348,7 +348,8 @@ def test_cross_val_score_with_score_func_regression():
     # Mean squared error; this is a loss function, so "scores" are negative
-    mse_scores = cross_val_score(reg, X, y, cv=5, scoring="mean_squared_error")
+    mse_scores = cross_val_score(reg, X, y, cv=5,
+                                 scoring="neg_mean_squared_error")
```
lease rename to neg_mse_scores or multiply by -1.

betatim committed:
Keep old metric names for deprecation period

Maintain old names during the deprecation period and update
tests to use better variables names.

betatim committed:
Keep old metric names in tests

ogrisel commented:
There is a failed doctest in `model_evaluation.rst`:

https://travis-ci.org/scikit-learn/scikit-learn/jobs/155644998#L3998

ogrisel commented:
Also line 204 of `get_scorer` should also filter out the deprecated scores from the list of valid options in its error message.

betatim committed:
Do not show deprecated scorers in list of valid names

betatim commented:
Thanks. I had updated the expected output but not the actual code ... both fixed now.

amueller commented on sklearn/metrics/scorer.py:
```diff
@@ -333,19 +362,27 @@ def make_scorer(score_func, greater_is_better=True, needs_proba=False,
                mean_squared_error=mean_squared_error_scorer,
                accuracy=accuracy_scorer, roc_auc=roc_auc_scorer,
                average_precision=average_precision_scorer,
-               log_loss=log_loss_scorer,
```
we can't remove that, right? It's public API.

betatim replied:
Correct. Wasn't paying enough attention apparently :(

amueller commented:
thanks for working on this, it's very important I think.

jnothman commented:
Sorry I've not been attuned to this thread. As I've proposed before, I think all deprecation should be in `get_scorer`: that's where the warning should be raised. The scorers themselves are not public API. Is that mistaken/

amueller commented:
Hm the scorer objects don't have underscores in their names, and the SCORERS dict that contains the instances is mentioned in the docs. I considered them public API.

jnothman commented:
`SCORERS` yes, but I think that was a bad design choice :) if we deprecate `SCORERS` at the same time, do we win? I don't think the scorers themselves were ever intended as public. They were never included in `classes.rst`, they are as useful as their names, and they should be immutable. In practice we have lots of names imported or declared in non-underscored modules that are not assumed to be public; these are mere module-level variables. Next you'll argue that `sklearn.metrics.scorer.qualified_name` is also public API!

betatim committed:
Do not remove log_loss scorer during deprecation period

betatim committed:
Distinguish instances from functions in deprecated

Treat (callable) instances different from functions in
the deprecated decorator

betatim commented:
My original implementation was via `get_scorer` (ef64969) but after chatting with @ogrisel we decided that people will have been "naughty" and used both the `sklearn.metrics.scorer.qualified_name` and `SCORERS` from the outside (and that they should get a deprecation warning).

I think if we can organise for those users to get a warning as well without too much acrobatics then we should. No opinion on whether that warning should be the rename deprecation warning or a "stop using these internals" warning.

GaelVaroquaux commented on sklearn/metrics/scorer.py:
```diff
@@ -33,14 +34,15 @@
 
 class _BaseScorer(six.with_metaclass(ABCMeta, object)):
-    def __init__(self, score_func, sign, kwargs):
+    def __init__(self, score_func, sign, kwargs, deprecation_msg=None):
```
I think that we will want to get rid of the "deprecation_msg" in two release. Maybe add a comment here saying that this should be removed for 0.20.

ogrisel replied:
Maybe we could rename that parameter to `_deprecation_msg` to make it explicit that this is not public API.

ogrisel replied:
Or even not put `_deprecation_msg` in the `__init__` at all but just as `None` initialized attribute on the `_BaseScorer` class and manually use 

```python
mean_scorer_error._deprecation_msg = "the message"
```

to activate the deprecation.

GaelVaroquaux commented on sklearn/metrics/scorer.py:
```diff
@@ -33,14 +34,15 @@
     def __call__(self, estimator, X, y, sample_weight=None):
-        pass
+        if self._deprecation_msg is not None:
+            warnings.warn(self._deprecation_msg, category=DeprecationWarning)
```
Don't you want a stacklevel=2 at least here, maybe even 3 (given that this method will always be called inside another)

GaelVaroquaux commented on sklearn/utils/deprecation.py:
```diff
@@ -76,6 +79,22 @@ def wrapped(*args, **kwargs):
 
+    def _decorate_instance(self, instance):
+        cls = instance.__class__
+        msg = "Thing %s is deprecated" % cls.__name__
```
Why "Thing %s is deprecated"? Why not "%s is deprecated"?

betatim replied:
Because I am not at all sure that we want to use `_decorate_instance`, wanted to put it out there as an idea: https://github.com/scikit-learn/scikit-learn/pull/7261#discussion_r76738685

(conclusion: discussion in line comments isn't very findable)

GaelVaroquaux commented:
I've finished my pass. I am :+1: with this PR once my small remarks are addressed.

Also, flake8 is unhappy (check travis). I guess that you should address that.

betatim commented:
Considering we are going back to discussing the basic approach I went from MRG -> WIP again.

To keep things on track a bit:

* should the warning be just for users via `get_scorer` or also others who "reach into" the module directly?
* if we want to warn everyone should we implement some addition to `utils.deprecated` so it can support the `make_scorer` use case or should we handle it via a argument to `make_scorer` that will be removed again in 0.20?

I am in favour of "warn everyone" and adding to `utils.deprecated`. Others with opinions?

jnothman commented on sklearn/utils/deprecation.py:
```diff
@@ -76,6 +79,22 @@ def wrapped(*args, **kwargs):
+            return instance(*args, **kwargs)
+
+        wrapped.__name__ = cls.__name__
+        wrapped.__dict__ = instance.__dict__
```
Surely this isn't enough to make it behave like an object of that class in cases other than `__call__`. I don't see how `_decorate_instance` is applicable in the general case, and hence why we're writing something general rather than simply using a mixin or something.

amueller commented:
we could do ``scorer.__call__ = deprecated("text")(scorer.__call__)``, right? Though the message would be weird? I would really rather deprecate the ``__call__`` method to cover all bases.
Maybe we can add an option to ``deprecated`` to overwrite the whole string?

@jnothman I would argue that all module level constants that don't start with an underscore could be considered public. It's very hard for users to check what is public and what is not otherwise. We could make the whole scorer submodule private in the future and import the important stuff into metrics (which I think we're already doing). If there are many more places where things are not made private properly, we should rename/deprecate them, or we'll end up with the same mess as we did with utils.

jnothman commented:
I think that policy is unhelpful. it's standard in Python, for instance to perform `from collections import defaultdict` but not expect anyone to import that from your module, despite not having done `from collections import defaultdict as _defaultdict`. I think we would be better off having a way to ensure that `classes.rst` is up-to-date.

jnothman commented:
That is, `classes.rst` should explicitly correspond to the idea of public API, and should be equivalent to the sum-total of `__all__` definitions.

lesteve commented:
The Travis failure comes from flake8 problems.

betatim commented:
> The Travis failure comes from flake8 problems, can you fix them?

Yes, just waiting to see if we can decide on how to do the deprecation, then implement that and take care of the `flake8` stuff.

ogrisel commented:
Unfortunately the SCORERS dict is already effectively part of the public API of scikit-learn 0.17.1:

```python
>>> import sklearn
>>> print(sklearn.__version__)
0.17.1
>>> from sklearn.metrics import SCORERS
>>> 'mean_squared_error' in SCORERS
True
```

We don't want to break our users code, so we have to put the deprecation in the scorer object's `__call__` method.

jnothman commented:
Yes, yes, I realise reality is against me. Go ahead. Make ugly code as
necessary!

On 31 August 2016 at 21:59, Olivier Grisel <notifications@github.com> wrote:

> Unfortunately the SCORERS dict is already effectively part of the public
> API of scikit-learn 0.17.1:
>
> >>> import sklearn>>> print(sklearn.__version__)0.17.1>>> from sklearn.metrics import SCORERS>>> 'mean_squared_error' in SCORERSTrue
>
> We don't want to break our users code, so we have to put the deprecation
> in the scorer object's __call__ method.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7261#issuecomment-243742313>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz670HyVRXa0QTKS1ZCLNp5O983J16ks5qlWyPgaJpZM4JuqCf>
> .
>

ogrisel commented:
Alternatively we could wrap the `__getitem__` method of the `SCORERS` instance and consider that `sklearn.metrics.scores` attributes are not part of the public API. That's slightly less ugly code.

But I think I am -0 on that and prefer to implement the uglier but fool proof version currently implemented in this PR.

lesteve commented:
It would be nice to add tests to make sure that accessing the old scorers (via the different ways mentioned above) does give DeprecationWarning as expected.

betatim committed:
Go back to warning all callers

ogrisel commented on sklearn/metrics/scorer.py:
```diff
@@ -34,15 +34,19 @@
-            warnings.warn(self._deprecation_msg, category=DeprecationWarning)
+            warnings.warn(self._deprecation_msg,
+                          category=DeprecationWarning,
+                          stacklevel=13)
```
Would it be possible to check that the `stacklevel=13` yield informative output in a test? I have no idea how this works and if this is testable or not.

betatim replied:
I will add a test as suggested https://github.com/scikit-learn/scikit-learn/pull/7261#issuecomment-243758677 and see what we see. For a simple script `stacklevel=13` refers to the right line where I made the choice to use an old name. There are lots of `Parallel` related things on the stack :-/ but now idea how well this works in general.

betatim commented:
@amueller can you work out why `scorer.__call__ = deprecated("text")(scorer.__call__)` doesn't print a message when it is called? I must be overlooking something but in a small test I made I couldn't get it to print anything. I used:
```python
from sklearn.ensemble import RandomForestClassifier; from sklearn.datasets import make_classification; from sklearn.model_selection import cross_val_score

from sklearn.metrics import SCORERS
from sklearn.metrics import get_scorer
from sklearn.metrics.scorer import mean_squared_error_scorer

clf = RandomForestClassifier(); X, y = make_classification()

print(cross_val_score(clf, X, y, scoring='mean_squared_error'))
print(cross_val_score(clf, X, y, scoring=SCORERS['mean_squared_error']))
print(cross_val_score(clf, X, y, scoring=get_scorer('mean_squared_error')))
print(cross_val_score(clf, X, y, scoring=mean_squared_error_scorer))
```

For the moment I went with https://github.com/scikit-learn/scikit-learn/pull/7261#discussion_r76979585

stack level: To get the warning to refer to the place where the user made the choice you have to go to something insane like 13??

betatim committed:
Remove unused import

amueller commented:
it seems you can't overwrite ``__call__`` on an object ?! doing ``object(argument)`` is equivalent to ``Class.__call__(object, argument)`` not ``object.__call__(object, argument)``. really?

amueller commented:
why WIP? LGTM

betatim commented:
MRG 😃

MechCoder commented:
Is this expected behaviour that just doing `get_scorer("mean_squared_error")` or `SCORERS["mean_squared_error"]` does not raise a deprecation warning as well?

MechCoder commented on doc/modules/model_evaluation.rst:
```diff
@@ -49,36 +49,36 @@ Common cases: predefined values
-and mean_squared_error, which measure the distance between the model
-and the data, are negated.
+than lower return values**.  Thus metrics which measure the distance between
+the model and the data, like mean_squared_error, are implemented as
```
model -> predictions?

nelson-liu replied:
should you link `mean_squared_error` and `neg_mean_squared_error` (below) to their associated functions?

betatim replied:
Here `mean_squared_error` refers to the metric and `neg_mean_squared_error` to the scorer. So I linked `mean_squared_error` to the metric, I think it makes it clearer, thansk for the tip!

I think you measure the quality of the model (via the predictions), so I vote to stick with what was already there.

amueller commented:
It's expected. I guess we could add deprecations to to ``get_scorer`` but I feel the case that someone gets it and doesn't call it is probably rare.

MechCoder commented:
Sorry, I don't understand. Do you mean calling `get_scorer` is rare?

jnothman commented:
He means calling y = get_scorer(x) and not calling y(...) is rare.

On 1 September 2016 at 12:51, Manoj Kumar <notifications@github.com> wrote:

> Sorry, I don't understand. Do you mean calling get_scorer is rare?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7261#issuecomment-243961650>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz67flzie7YJhkf1ENUmdnbCkrHDxvks5qlj2igaJpZM4JuqCf>
> .
>

MechCoder commented:
hmm ok.

jnothman commented:
I HAVE A GRAND SOLUTION: move sklearn/metrics/scorer.py to
sklearn/model_selection where it belongs. Deprecate everything in the old
module and only make get_scorer and list_scorers (as yet uncreated) public!

I'm only half kidding, too.

On 1 September 2016 at 13:16, Manoj Kumar <notifications@github.com> wrote:

> hmm ok.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7261#issuecomment-243964756>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz62Yhgppb6sfiiKrNVKViRFOqUUWLks5qlkOJgaJpZM4JuqCf>
> .
>

MechCoder commented:
And have Google fund it as part as of a Google Summer of Code project? (sorry @raghavrv could not help it. :p )

betatim committed:
Link to mean_squared_error function

Makes the distinction between the metric and the scorer a bit
easier for the reader.

betatim commented:
@MechCoder and @nelson-liu addressed your documentation comment.

> And have Google fund it as part as of a Google Summer of Code project?

Think I am too old to qualify for GSOC anymore 😢

GaelVaroquaux commented:
With regards to moving everything, maybe? I am not sure that it is worth the cost. And if we do it, we should do it now!

With regard to @amueller 's suggestion to decorate the __call__, I am -1 on it: these things tend to break documentation and pickling (pickling is important for parallel computing).

Aside from that, I am +1 for merge, and I think that I saw LGTM from @amueller, so we can merge. Waiting a little for discussion on the moving everything around, though I am -1.

jnothman commented:
No, I don't think moving scorers is worth the cost. We're not making any
fundamental changes. The one advantage might be that it makes a clearer
distinction between metrics and scorers, which users sometimes have a
problem following.

On 1 September 2016 at 16:39, Gael Varoquaux <notifications@github.com>
wrote:

> With regards to moving everything, maybe? I am not sure that it is worth
> the cost. And if we do it, we should do it now!
>
> With regard to @amueller <https://github.com/amueller> 's suggestion to
> decorate the *call*, I am -1 on it: these things tend to break
> documentation and pickling (pickling is important for parallel computing).
>
> Aside from that, I am +1 for merge, and I think that I saw LGTM from
> @amueller <https://github.com/amueller>, so we can merge. Waiting a
> little for discussion on the moving everything around, though I am -1.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7261#issuecomment-243989683>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6--ja-gVF-YBGFqUUUQIJIbpgNCaks5qlnM_gaJpZM4JuqCf>
> .
>

jnothman commented on doc/modules/model_evaluation.rst:
```diff
@@ -49,36 +49,37 @@ Common cases: predefined values
-and the data, are negated.
+than lower return values**.  Thus metrics which measure the distance between
+the model and the data, like :func:`metrics.mean_squared_error`, are
+implemented as neg_mean_squared_error which return the negated value
```
"implemented" -> "available" or "catalogued" or "plugged in"

jnothman commented on sklearn/metrics/scorer.py:
```diff
@@ -313,12 +327,31 @@ def make_scorer(score_func, greater_is_better=True, needs_proba=False,
+                   'be removed in 0.20.')
 mean_squared_error_scorer = make_scorer(mean_squared_error,
                                         greater_is_better=False)
+mean_squared_error_scorer._deprecation_msg = deprecation_msg
```
I find this weird relative to a method or something on the object, but whatever; it's all private internals of private objects.

jnothman commented on sklearn/metrics/scorer.py:
```diff
@@ -191,9 +203,11 @@ def get_scorer(scoring):
         try:
             scorer = SCORERS[scoring]
         except KeyError:
+            scorers = [scorer for scorer in SCORERS
```
I'd actually like a `list_scorers` function, but it's low priority.

jnothman replied:
and `SCORERS` deprecated from the public API :)

betatim replied:
I take this as an invitation to create an issue for this instead of fixing it here ok? 😁

jnothman commented on sklearn/metrics/scorer.py:
```diff
@@ -37,10 +38,15 @@ def __init__(self, score_func, sign, kwargs):
+        if self._deprecation_msg is not None:
+            warnings.warn(self._deprecation_msg,
+                          category=DeprecationWarning,
+                          stacklevel=13)
```
13???

jnothman replied:
This is going to depend on the invocation, and unless you inspect the stack trace and find the right place... Just use a small stacklevel. Warnings suck.

betatim replied:
What is a good default then? If you use `cross_val_score()` you need stackelvel=13 because of all the `Parallel` stuff. So I have no opinion really on what to set it to. It seems like getting the warning to actually point to where you made the choice that leads to the warning is nearly impossible??

GaelVaroquaux replied:
> What is a good default then?

Well, if a scorer is called, what level is needed to have the warning
displayed where the scorer is called?

betatim replied:
For this (or use via `cross_val_score`)
```
scorer = get_scorer('mean_squared_error')
clf.fit(X, y)
scorer(clf, X, y)
```
you need `stacklevel=2` which will point you at `scikit-learn/sklearn/metrics/scorer.py:89`. For the `cross_val_score` you can go higher but then you don't get a warning for the case where you call it directly.

Neither location is particularly informative for the user that didn't expect to get the warning I'd say. either you get that you made this choice somewhere in your code, or you don't get it :-/

jnothman replied:
Yes, it's always very hard to get informative line numbers for the user from warnings. I don't know why they don't store (at least, without printing) a traceback.

betatim committed:
Small documentation fix

betatim commented:
Addressed the documentation comment and stacklevel.

jnothman commented on sklearn/metrics/scorer.py:
```diff
@@ -37,10 +38,15 @@ def __init__(self, score_func, sign, kwargs):
+        # XXX deprecation_msg property again and make __call__ abstract again
+        self._deprecation_msg = None
 
-    @abstractmethod
```
I think you can keep this.

betatim committed:
Keep abstractmethod decorator

jnothman commented on sklearn/metrics/scorer.py:
```diff
@@ -42,6 +42,7 @@ def __init__(self, score_func, sign, kwargs):
         # XXX deprecation_msg property again and make __call__ abstract again
```
comment update?

betatim commented:
*bump*

jnothman commented:
:)

jnothman commented:
Examples not yet updated?
* `examples/model_selection/plot_underfitting_overfitting.py`
* `examples/plot_kernel_ridge_regression.py`

ogrisel commented:
The fix it-self looks good to me. +1 for merge once the examples are updated.

betatim committed:
Update examples to new scorer names

betatim commented:
Updated.

jnothman commented on examples/plot_kernel_ridge_regression.py:
```diff
@@ -154,14 +154,14 @@
+                   scoring="neg_mean_squared_error", cv=10)
 
-plt.plot(train_sizes, test_scores_svr.mean(1), 'o-', color="r",
+plt.plot(train_sizes, -test_scores_svr.mean(1), 'o-', color="r",
```
I think you should leave off this negation and change the ylabel to say "Negative".

jnothman replied:
But what you have here is okay; better than what's currently plotted at http://scikit-learn.org/stable/auto_examples/plot_kernel_ridge_regression.html.

@ogrisel, any preference for plotting -MSE vs plotting MSE?

betatim replied:
I find plotting MSE and seeing it decrease easier on my brain than having to think "error, smaller is better but here it is negated so going up means closer to zero ..."

jnothman replied:
The question from my perspective is whether it's worth encouraging users to negate output. Or whether, regardless of scoring function we'd like to encourage the convention that better is up.

betatim replied:
I have no opinion on this beyond previous comment. Who is going to arbitrate/decide?

amueller replied:
No strong opinion here. I'm fine with the ``-``

jnothman replied:
Merge then?

jnothman commented:
LGTM other than that too

lesteve commented:
> It would be nice to add tests to make sure that accessing the old scorers (via the different ways mentioned above) does give DeprecationWarning as expected.

@betatim bonus points if you feel like tackling my comment from a week ago which got lost in other (very likely more important) considerations.

betatim committed:
Test that warning is raised when using deprecated scorer

betatim commented:
Tested.

I skipped testing the `from sklearn.metrics.scorer import mean_squared_error_scorer` kind of access. I think it would require a bit of import acrobatics to do based on a string name and not sure that is worth it.

amueller commented:
merge?

betatim commented:
Make it so.

GaelVaroquaux commented:
Yey!

(I didn't do much, but I pressed the green button. What a pleasure!)

GaelVaroquaux commented:
Thanks heaps @betatim !

amueller commented:
awesome, thanks! this was was overdue

betatim commented:
Thanks a lot for all the comments and extra 👀!

raghavrv commented:
Yay :beers: Much needed PR!

raghavrv commented:
@jnothman This also closes #6028 #5023 right?

amueller commented:
It does. Closed those. Thanks @raghavrv
